File: client/app/components/training-mode/ExerciseLogCard.js
"use client";

import { useEffect, useRef, useState } from "react";
import { useTranslations } from "next-intl";
import { LineChart as ProgressIcon, BookOpen, NotebookPen, Check } from "lucide-react";
import { Chip } from "@heroui/react/chip";
import { Button } from "@heroui/react/button";
import { Tooltip } from "@heroui/react/tooltip";
import { TextField } from "@heroui/react/textfield";
import { Input } from "@heroui/react/input";
import ExerciseVideoPlayer from "./ExerciseVideoPlayer";
import ExerciseNotesModal from "./ExerciseNotesModal";
import ExerciseInstructionsModal from "./ExerciseInstructionsModal";
import ClientExerciseInsightsModal from "./ClientExerciseInsightsModal";
import NewFeatureTooltip from "@/app/components/NewFeatureTooltip";
import { hasTempoValue, hasRirValue, formatClock } from "@/utils/workout";
import { categoryOf, loggedFieldsFor, targetOnlyFieldsFor } from "@/utils/exerciseTrackingTypes";

// Precomputed Tailwind arbitrary-value grid templates — kept as whole literal
// strings (rather than built dynamically) so Tailwind's JIT scanner can find
// them. Every editable/previous column is "1fr"; read-only target columns
// (tempo/rir/rpe, Sets & Reps only) get a compact fixed width; 24px/28px for
// the set index/done tick. Sets & Reps has a fixed 2-column editable set
// (weight, reps) with 0-3 target columns depending on the coach's
// checklist; Time-Based has 0-4 editable columns and no targets.
const SET_ROW_COLS_SETS_REPS_BY_TARGET_COUNT = [
    "grid-cols-[24px_1fr_1fr_1fr_28px]",                     // 0 targets
    "grid-cols-[24px_1fr_1fr_1fr_36px_28px]",                // 1
    "grid-cols-[24px_1fr_1fr_1fr_36px_36px_28px]",           // 2
    "grid-cols-[24px_1fr_1fr_1fr_36px_36px_36px_28px]",      // 3 (tempo + rir + rpe)
];
const SET_ROW_COLS_TIME_BASED_BY_COUNT = [
    "grid-cols-[24px_1fr_28px]",             // 0: coach selected no metrics for this exercise (edge case)
    "grid-cols-[24px_1fr_1fr_28px]",         // 1
    "grid-cols-[24px_1fr_1fr_1fr_28px]",     // 2
    "grid-cols-[24px_1fr_1fr_1fr_1fr_28px]", // 3
    "grid-cols-[24px_1fr_1fr_1fr_1fr_1fr_28px]", // 4
];

const FIELD_LABEL_KEY = {
    weight: "weight",
    reps: "repsShort",
    tempo: "tempo",
    rir: "rir",
    rpe: "rpe",
    duration_seconds: "duration",
    distance_km: "distance",
    incline_percent: "incline",
    speed_kmh: "speed",
};

function hasFieldValue(v) {
    return v !== null && v !== undefined && v !== "";
}

// rest_seconds is prescribed (a base field, always) but never a displayed
// column here — it drives the rest timer elsewhere in session/page.js, not
// a table cell. targetOnlyFieldsFor() doesn't know that distinction, so it's
// filtered out at the two call sites below.
function displayTargetFields(exercise) {
    return targetOnlyFieldsFor(exercise).filter((field) => field !== "rest_seconds");
}

// tempo's "not filled in" sentinel is the literal string "-" (see
// useTrainingPlan.js's defaultSetFor) — hasFieldValue alone would treat that
// as a real value. rir has its own similar historical helper; rpe has no
// sentinel, so hasFieldValue is exact for it.
function hasTargetValue(field, value) {
    if (field === "tempo") return hasTempoValue(value);
    if (field === "rir") return hasRirValue(value);
    return hasFieldValue(value);
}

// One exercise within a Training Mode session: a lazy video, per-set targets
// shown alongside each row, and an editable grid of logged sets (previous ·
// weight · reps · done). Flat, borderless layout — exercises are separated by
// the session list's own spacing/dividers rather than a card container, so
// the video, header and table read as one continuous section (matches the
// mobile app's Training Mode, which deliberately dropped card chrome here).
export default function ExerciseLogCard({ exercise, previous, focusSetIndex, isFirstExercise, onChangeSet, onToggleSet, onChangeNote }) {
    const t = useTranslations("portal.training");
    const [insightsOpen, setInsightsOpen] = useState(false);
    const [instructionsOpen, setInstructionsOpen] = useState(false);
    const [notesOpen, setNotesOpen] = useState(false);
    const weightRefs = useRef([]);

    const prescribed = exercise.prescribed ?? [];
    const hasNote    = Boolean(exercise.note?.trim());
    const category   = categoryOf(exercise);
    // Column visibility is entirely the coach's choice now (tracked_metrics,
    // set in the exercise library) — not whether any particular set happens
    // to have a value yet. editableFields = what the client types in;
    // targetFields = read-only, from the coach's prescription (tempo/rir/rpe
    // for Sets & Reps only; Time-Based's metrics are editable, not targets).
    const editableFields = loggedFieldsFor(exercise);
    const targetFields   = displayTargetFields(exercise);
    const rowCols = category === "sets_reps"
        ? SET_ROW_COLS_SETS_REPS_BY_TARGET_COUNT[targetFields.length]
        : SET_ROW_COLS_TIME_BASED_BY_COUNT[editableFields.length];

    // Moves keyboard focus to the next set's weight field once the set before
    // it is marked done — lets a client log a whole exercise without reaching
    // for the mouse. Mirrors the mobile app's focusSetIndex behavior.
    useEffect(() => {
        if (focusSetIndex == null) return;
        weightRefs.current[focusSetIndex]?.focus();
    }, [focusSetIndex]);

    function previousFor(setOrder) {
        const match = (previous ?? []).find(p => p.set_order === setOrder);
        if (!match) return null;
        if (category === "sets_reps") {
            if (match.weight == null || match.reps == null) return null;
            return `${match.weight}kg × ${match.reps}`;
        }
        // time_based
        const parts = [];
        if (match.duration_seconds != null) parts.push(formatClock(match.duration_seconds));
        if (match.distance_km != null) parts.push(`${match.distance_km}km`);
        return parts.length > 0 ? parts.join(" · ") : null;
    }

    // Weight has no prescribed counterpart (training_sets never had a weight
    // column) — its placeholder stays "what you did last time". Every other
    // logged field IS something the coach set a target for on this set, so
    // showing that target as the placeholder tells the client what to aim
    // for, same as reps' targetReps below.
    function placeholderFor(field, setOrder, sIdx) {
        if (field === "weight") {
            const match = (previous ?? []).find(p => p.set_order === setOrder);
            return match?.weight != null ? `${match.weight}kg` : "kg";
        }
        const target = prescribed[sIdx]?.[field];
        if (!hasFieldValue(target)) return "—";
        if (field === "duration_seconds") return formatClock(target);
        if (field === "distance_km") return `${target}km`;
        if (field === "incline_percent") return `${target}%`;
        if (field === "speed_kmh") return `${target}km/h`;
        return String(target);
    }

    const metaParts = [exercise.muscle_group, exercise.equipment].filter(Boolean);

    return (
        <div className="flex flex-col">
            {(exercise.youtube_url || exercise.video_path || exercise.thumbnail_path) && (
                <div className="mb-3">
                    <ExerciseVideoPlayer
                        youtubeUrl={exercise.youtube_url}
                        videoPath={exercise.video_path}
                        thumbnailPath={exercise.thumbnail_path}
                        name={exercise.name}
                        watchLabel={t("watchVideo")}
                        watchOnYoutubeLabel={t("watchOnYoutube")}
                    />
                </div>
            )}

            {/* Header */}
            <div className="flex items-start gap-1">
                <div className="flex-1 min-w-0">
                    <p className="text-sm font-semibold text-foreground">{exercise.name}</p>
                    {metaParts.length > 0 && (
                        <div className="flex items-center gap-1.5 flex-wrap mt-1">
                            {metaParts.map(part => (
                                <Chip key={part} size="sm" variant="soft">{part}</Chip>
                            ))}
                        </div>
                    )}
                </div>
                <Tooltip>
                    <Button
                        isIconOnly
                        variant="ghost"
                        size="sm"
                        aria-label={t("progress")}
                        onClick={() => setInsightsOpen(true)}
                        className="shrink-0 text-muted-foreground"
                    >
                        <ProgressIcon className="w-4 h-4" />
                    </Button>
                    <Tooltip.Content>{t("progress")}</Tooltip.Content>
                </Tooltip>
                <span title={t("instructions")}>
                    <NewFeatureTooltip
                        featureKey="exercise_instructions_hint"
                        active={!!isFirstExercise}
                        message={t("instructionsHint")}
                        dismissLabel={t("instructionsHintDismiss")}
                        badgeLabel={t("instructionsNewFeature")}
                        onTriggerClick={() => setInstructionsOpen(true)}
                        triggerClassName="shrink-0 w-8 h-8 flex items-center justify-center rounded-full text-muted-foreground hover:bg-default hover:text-foreground transition-colors cursor-pointer"
                    >
                        <BookOpen className="w-4 h-4" />
                    </NewFeatureTooltip>
                </span>
                <Tooltip>
                    <Button
                        isIconOnly
                        variant="ghost"
                        size="sm"
                        aria-label={t("notes")}
                        onClick={() => setNotesOpen(true)}
                        className={`shrink-0 ${hasNote ? "text-primary" : "text-muted-foreground"}`}
                    >
                        <NotebookPen className="w-4 h-4" />
                    </Button>
                    <Tooltip.Content>{t("notes")}</Tooltip.Content>
                </Tooltip>
            </div>

            {exercise.notes && (
                <div className="border-s-2 border-amber-500/60 ps-3 mt-2">
                    <p dir="auto" className="text-xs text-muted-foreground whitespace-pre-wrap">{exercise.notes}</p>
                </div>
            )}

            {/* Sets */}
            <div className="flex flex-col mt-3">
                <div className={`grid ${rowCols} gap-1.5 text-[10px] font-semibold uppercase tracking-wide text-muted-foreground/50 pb-1`}>
                    <span className="text-center truncate">{t("set")}</span>
                    <span className="text-center truncate">{t("previous")}</span>
                    {editableFields.map((field) => (
                        <span key={field} className="text-center truncate">{t(FIELD_LABEL_KEY[field])}</span>
                    ))}
                    {targetFields.map((field) => (
                        <span key={field} className="text-center truncate">{t(FIELD_LABEL_KEY[field])}</span>
                    ))}
                    <Check className="w-3.5 h-3.5 mx-auto text-muted-foreground/50" />
                </div>

                <div className="flex flex-col divide-y divide-border/30">
                    {exercise.sets.map((set, sIdx) => {
                        const targetReps = prescribed[sIdx]?.reps ?? null;
                        return (
                            <div
                                key={sIdx}
                                className={`grid ${rowCols} gap-1.5 items-center py-1 rounded-lg transition-colors ${set.completed ? "bg-success/10" : ""}`}
                            >
                                <span className="text-xs text-muted-foreground/60 text-center">{sIdx + 1}</span>
                                <span className="text-[10px] text-muted-foreground/50 text-center truncate">{previousFor(set.set_order) ?? "—"}</span>

                                {editableFields.map((field, fIdx) => (
                                    <TextField key={field} value={set[field] ?? ""} onChange={(v) => onChangeSet(sIdx, field, v)} aria-label={t(FIELD_LABEL_KEY[field])} variant="secondary" fullWidth>
                                        <Input
                                            ref={fIdx === 0 ? (el) => { weightRefs.current[sIdx] = el; } : undefined}
                                            inputMode={field === "reps" || field === "duration_seconds" ? "numeric" : "decimal"}
                                            placeholder={field === "reps" ? (targetReps ?? "—") : placeholderFor(field, set.set_order, sIdx)}
                                            className={`text-center px-1 py-1 text-xs ${set.completed ? "border-transparent bg-transparent shadow-none" : ""}`}
                                        />
                                    </TextField>
                                ))}

                                {targetFields.map((field) => {
                                    const raw = prescribed[sIdx]?.[field];
                                    return (
                                        <span key={field} className="text-xs text-muted-foreground text-center truncate">
                                            {hasTargetValue(field, raw) ? raw : "—"}
                                        </span>
                                    );
                                })}

                                <button
                                    type="button"
                                    onClick={() => onToggleSet(sIdx)}
                                    aria-label={t("markDone")}
                                    aria-pressed={set.completed}
                                    className={`mx-auto flex items-center justify-center w-7 h-7 rounded-lg border-2 transition-colors cursor-pointer ${
                                        set.completed ? "bg-success border-success text-white" : "border-muted-foreground/40 text-transparent hover:border-muted-foreground/70"
                                    }`}
                                >
                                    <Check className="w-4 h-4" strokeWidth={2.5} />
                                </button>
                            </div>
                        );
                    })}
                </div>
            </div>

            <ClientExerciseInsightsModal
                open={insightsOpen}
                onClose={() => setInsightsOpen(false)}
                exercise={exercise}
            />
            <ExerciseInstructionsModal
                open={instructionsOpen}
                onClose={() => setInstructionsOpen(false)}
                exercise={exercise}
            />
            <ExerciseNotesModal
                open={notesOpen}
                onClose={() => setNotesOpen(false)}
                initialValue={exercise.note}
                onSave={onChangeNote}
                exercise={exercise}
            />
        </div>
    );
}

File: mobile/lib/features/training/widgets/exercise_log_card.dart
import 'package:flutter/material.dart';

import '../../../core/theme/app_theme.dart';
import '../../../l10n/generated/app_localizations.dart';
import '../../../shared/models/workout_log.dart';
import '../../../shared/models/workout_session.dart';
import '../../../shared/utils/exercise_tracking_types.dart' as tracking;
import '../../../shared/utils/format_amount.dart';
import '../../../shared/utils/localization.dart';
import '../../../shared/utils/workout.dart';
import 'coach_note_modal.dart';
import 'exercise_insights_modal.dart';
import 'exercise_notes_modal.dart';
import 'exercise_video.dart';

/// Reads one named field off a [PrescribedSet] — the coach's target, shown
/// either as a read-only column (tempo/rir/rpe) or as an editable field's
/// placeholder (duration/distance/incline/speed, reps).
String? _prescribedValue(PrescribedSet? target, String field) {
  if (target == null) return null;
  return switch (field) {
    'reps' => target.reps,
    'tempo' => target.tempo,
    'rir' => target.rir,
    'rpe' => target.rpe,
    'duration_seconds' => target.durationSeconds,
    'distance_km' => target.distanceKm,
    'incline_percent' => target.inclinePercent,
    'speed_kmh' => target.speedKmh,
    _ => null,
  };
}

/// Reads one named editable field off a [SessionSet] being logged.
String _setValue(SessionSet set, String field) => switch (field) {
      'weight' => set.weight,
      'reps' => set.reps,
      'duration_seconds' => set.durationSeconds,
      'distance_km' => set.distanceKm,
      'incline_percent' => set.inclinePercent,
      'speed_kmh' => set.speedKmh,
      _ => '',
    };

/// Weight has no prescribed counterpart — its placeholder stays "what the
/// client did last time". Every other editable field IS something the coach
/// set a target for on this set, so showing that target as the placeholder
/// tells the client what to aim for.
String _placeholderFor(
  String field,
  PrescribedSet? target,
  PreviousSet? previous,
) {
  if (field == 'weight') {
    return previous?.weight != null ? '${prettyAmount(previous!.weight!)}kg' : 'kg';
  }
  final raw = _prescribedValue(target, field);
  if (raw == null || raw.isEmpty) return '—';
  if (field == 'duration_seconds') {
    final secs = int.tryParse(raw);
    return secs != null ? formatClock(secs) : raw;
  }
  if (field == 'distance_km') return '${raw}km';
  if (field == 'incline_percent') return '$raw%';
  if (field == 'speed_kmh') return '${raw}km/h';
  return raw;
}

/// One exercise within a Training Mode session: a lazy video, per-set targets
/// shown alongside each row, and an editable grid of logged sets (previous ·
/// [logged fields] · [target fields] · done). Which fields are editable vs.
/// read-only targets is entirely the coach's choice (tracking_type +
/// tracked_metrics) — see shared/utils/exercise_tracking_types.dart.
class ExerciseLogCard extends StatefulWidget {
  const ExerciseLogCard({
    super.key,
    required this.exercise,
    required this.previous,
    required this.onChangeSet,
    required this.onToggleSet,
    required this.onChangeNote,
    required this.videoUrl,
    this.focusSetIndex,
  });

  final SessionExercise exercise;
  final List<PreviousSet> previous;
  final String? videoUrl;

  /// Index of the set whose first editable field should take keyboard focus
  /// once the set before it is marked done. Null outside of that moment.
  final int? focusSetIndex;
  final void Function(int setIndex, String field, String value) onChangeSet;
  final void Function(int setIndex) onToggleSet;
  final ValueChanged<String> onChangeNote;

  @override
  State<ExerciseLogCard> createState() => _ExerciseLogCardState();
}

class _ExerciseLogCardState extends State<ExerciseLogCard> {
  late List<FocusNode> _weightFocusNodes =
      List.generate(widget.exercise.sets.length, (_) => FocusNode());

  @override
  void didUpdateWidget(covariant ExerciseLogCard oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (widget.exercise.sets.length != _weightFocusNodes.length) {
      for (final node in _weightFocusNodes) {
        node.dispose();
      }
      _weightFocusNodes =
          List.generate(widget.exercise.sets.length, (_) => FocusNode());
    }

    final target = widget.focusSetIndex;
    if (target != null &&
        target != oldWidget.focusSetIndex &&
        target >= 0 &&
        target < _weightFocusNodes.length) {
      WidgetsBinding.instance.addPostFrameCallback((_) {
        if (mounted) _weightFocusNodes[target].requestFocus();
      });
    }
  }

  @override
  void dispose() {
    for (final node in _weightFocusNodes) {
      node.dispose();
    }
    super.dispose();
  }

  bool get _hasVideo =>
      (widget.exercise.youtubeUrl ?? '').isNotEmpty ||
      (widget.videoUrl ?? '').isNotEmpty;

  bool get _hasNote => widget.exercise.note.trim().isNotEmpty;

  @override
  Widget build(BuildContext context) {
    final l10n = AppLocalizations.of(context);
    final muted = context.appColors.mutedForeground;
    final exercise = widget.exercise;
    final locale = Localizations.localeOf(context).languageCode;
    final exerciseName = localizedField(
      base: exercise.libraryNameEn ?? exercise.name,
      arabic: exercise.libraryNameAr,
      localeCode: locale,
    );
    final muscleGroup = localizedField(
      base: exercise.muscleGroup ?? '',
      arabic: exercise.muscleGroupAr,
      localeCode: locale,
    );
    final equipment = localizedField(
      base: exercise.equipment ?? '',
      arabic: exercise.equipmentAr,
      localeCode: locale,
    );

    final category = tracking.categoryOf(exercise.trackingType);
    // rest_seconds is prescribed (a base field, always) but never a
    // displayed column — it drives the rest timer, not a table cell.
    final editableFields = tracking
        .loggedFieldsFor(exercise.trackingType, exercise.trackedMetrics);
    final targetFields = tracking
        .targetOnlyFieldsFor(exercise.trackingType, exercise.trackedMetrics)
        .where((f) => f != 'rest_seconds')
        .toList();

    // Flat, borderless layout (Strong-style) — exercises are separated by the
    // list's own spacing/dividers rather than a card container, so the video,
    // header and table read as one continuous section.
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        if (_hasVideo) ...[
          ExerciseVideo(
            youtubeUrl: exercise.youtubeUrl,
            videoUrl: widget.videoUrl,
          ),
          const SizedBox(height: 12),
        ],

        // Header
        Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(exerciseName,
                      style: const TextStyle(
                          fontSize: 14, fontWeight: FontWeight.w600)),
                  const SizedBox(height: 4),
                  Wrap(
                    spacing: 6,
                    runSpacing: 4,
                    crossAxisAlignment: WrapCrossAlignment.center,
                    children: [
                      if (muscleGroup.isNotEmpty) _chip(context, muscleGroup),
                      if (equipment.isNotEmpty)
                        _chip(context, equipment,
                            color: const Color(0xFF8B5CF6)),
                    ],
                  ),
                ],
              ),
            ),
            _IconAction(
              icon: Icons.show_chart,
              tooltip: l10n.trainingInsights,
              onTap: () => showModalBottomSheet(
                context: context,
                isScrollControlled: true,
                useSafeArea: true,
                builder: (_) => ExerciseInsightsModal(
                  exerciseLibraryId: exercise.exerciseLibraryId,
                  exerciseId: exercise.exerciseId,
                  exerciseName: exerciseName,
                ),
              ),
            ),
            _IconAction(
              icon: Icons.menu_book_outlined,
              tooltip: l10n.trainingCoachNote,
              onTap: () => showModalBottomSheet(
                context: context,
                isScrollControlled: true,
                useSafeArea: true,
                builder: (_) => CoachNoteModal(
                  instructionsEn: exercise.instructionsEn,
                  instructionsAr: exercise.instructionsAr,
                ),
              ),
            ),
            _IconAction(
              icon: Icons.edit_note,
              tooltip: l10n.trainingExerciseNote,
              highlighted: _hasNote,
              onTap: () async {
                final result = await showModalBottomSheet<String>(
                  context: context,
                  isScrollControlled: true,
                  useSafeArea: true,
                  builder: (_) => ExerciseNotesModal(
                    initialValue: exercise.note,
                    exerciseLibraryId: exercise.exerciseLibraryId,
                    exerciseId: exercise.exerciseId,
                  ),
                );
                if (result != null) widget.onChangeNote(result);
              },
            ),
          ],
        ),
        const SizedBox(height: 12),

        // Column headers
        _SetRow.header(
          l10n: l10n,
          muted: muted,
          editableFields: editableFields,
          targetFields: targetFields,
        ),
        const SizedBox(height: 4),
        for (var i = 0; i < exercise.sets.length; i++)
          _SetRow(
            index: i,
            set: exercise.sets[i],
            previous: previousSetFor(widget.previous, exercise.sets[i].setOrder),
            target:
                i < exercise.prescribed.length ? exercise.prescribed[i] : null,
            category: category,
            editableFields: editableFields,
            targetFields: targetFields,
            weightFocusNode: _weightFocusNodes[i],
            onChange: (field, value) => widget.onChangeSet(i, field, value),
            onToggle: () => widget.onToggleSet(i),
          ),
      ],
    );
  }

  Widget _chip(BuildContext context, String label, {Color? color}) {
    final c = color ?? context.appColors.mutedForeground;
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
      decoration: BoxDecoration(
        color: c.withValues(alpha: 0.12),
        borderRadius: BorderRadius.circular(999),
      ),
      child: Text(label, style: TextStyle(fontSize: 11, color: c)),
    );
  }
}

class _IconAction extends StatelessWidget {
  const _IconAction({
    required this.icon,
    required this.tooltip,
    required this.onTap,
    this.highlighted = false,
  });

  final IconData icon;
  final String tooltip;
  final VoidCallback onTap;
  final bool highlighted;

  @override
  Widget build(BuildContext context) {
    final color = highlighted
        ? Theme.of(context).colorScheme.primary
        : context.appColors.mutedForeground;
    return Tooltip(
      message: tooltip,
      child: IconButton(
        onPressed: onTap,
        icon: Icon(icon, size: 18, color: color),
        visualDensity: VisualDensity.compact,
        padding: EdgeInsets.zero,
        constraints: const BoxConstraints(minWidth: 32, minHeight: 32),
      ),
    );
  }
}

/// A dynamic set row: # · previous · [editable logged fields] · [read-only
/// target fields] · done. Column set is entirely the coach's choice
/// (tracked_metrics), not whether a particular set happens to have a value.
/// Rest lives entirely in the floating [RestTimerBar] once a set completes —
/// it never occupies table space.
class _SetRow extends StatelessWidget {
  const _SetRow({
    required this.index,
    required this.set,
    required this.previous,
    required this.target,
    required this.category,
    required this.editableFields,
    required this.targetFields,
    required this.onChange,
    required this.onToggle,
    this.weightFocusNode,
  })  : isHeader = false,
        l10n = null,
        muted = null;

  const _SetRow.header({
    required this.l10n,
    required this.muted,
    required this.editableFields,
    required this.targetFields,
  })  : isHeader = true,
        index = 0,
        set = null,
        previous = null,
        target = null,
        category = tracking.setsReps,
        onChange = null,
        onToggle = null,
        weightFocusNode = null;

  final bool isHeader;
  final int index;
  final SessionSet? set;
  final PreviousSet? previous;
  final PrescribedSet? target;
  final String category;
  final List<String> editableFields;
  final List<String> targetFields;
  final void Function(String field, String value)? onChange;
  final VoidCallback? onToggle;
  final AppLocalizations? l10n;
  final Color? muted;

  /// Takes keyboard focus once the previous set is marked done — the row
  /// itself carries no visual "next up" treatment, just the moved focus.
  final FocusNode? weightFocusNode;

  @override
  Widget build(BuildContext context) {
    if (isHeader) {
      final style = TextStyle(
        fontSize: 10,
        fontWeight: FontWeight.w600,
        letterSpacing: 0.5,
        color: muted!.withValues(alpha: 0.7),
      );
      Widget header(String label, {double? width}) {
        final text = Text(label.toUpperCase(),
            style: style, textAlign: TextAlign.center, overflow: TextOverflow.ellipsis);
        return width != null
            ? SizedBox(width: width, child: text)
            : Expanded(child: text);
      }

      return Row(
        children: [
          header(l10n!.trainingSet, width: 24),
          const SizedBox(width: 4),
          header(l10n!.trainingPreviousShort),
          for (final field in editableFields)
            header(tracking.fieldLabel(l10n!, field)),
          for (final field in targetFields)
            header(tracking.fieldLabel(l10n!, field), width: 34),
          SizedBox(
            width: 32,
            child: Icon(Icons.check, size: 14, color: style.color),
          ),
        ],
      );
    }

    final s = set!;
    final mutedColor = context.appColors.mutedForeground;
    final success = context.appColors.success;
    final previousText = previousSetLabel(previous, category);

    return Container(
      margin: const EdgeInsets.symmetric(vertical: 2),
      padding: const EdgeInsets.symmetric(vertical: 1),
      decoration: BoxDecoration(
        color: s.completed ? success.withValues(alpha: 0.1) : null,
        borderRadius: BorderRadius.circular(8),
      ),
      child: Row(
        children: [
          SizedBox(
            width: 24,
            child: Text('${index + 1}',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 12, color: mutedColor)),
          ),
          const SizedBox(width: 4),
          Expanded(
            child: Text(previousText ?? '—',
                textAlign: TextAlign.center,
                style: TextStyle(
                    fontSize: 10,
                    fontWeight: FontWeight.w400,
                    color: mutedColor.withValues(alpha: 0.6)),
                overflow: TextOverflow.ellipsis),
          ),
          for (var i = 0; i < editableFields.length; i++)
            Expanded(
              child: _input(
                context,
                _setValue(s, editableFields[i]),
                _placeholderFor(editableFields[i], target, previous),
                editableFields[i],
                decimal: editableFields[i] != 'reps' &&
                    editableFields[i] != 'duration_seconds',
                focusNode: i == 0 ? weightFocusNode : null,
                plain: s.completed,
              ),
            ),
          for (final field in targetFields)
            SizedBox(
              width: 34,
              child: Text(
                _prescribedValue(target, field) ?? '—',
                textAlign: TextAlign.center,
                style: TextStyle(fontSize: 12, color: mutedColor),
              ),
            ),
          SizedBox(
            width: 32,
            child: Center(child: _doneButton(context, s.completed, success)),
          ),
        ],
      ),
    );
  }

  Widget _input(BuildContext context, String value, String hint, String field,
      {bool decimal = false, FocusNode? focusNode, bool plain = false}) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 2),
      child: TextFormField(
        initialValue: value,
        focusNode: focusNode,
        onChanged: (v) => onChange!(field, v),
        textAlign: TextAlign.center,
        keyboardType: TextInputType.numberWithOptions(
            decimal: decimal, signed: false),
        style: const TextStyle(fontSize: 13),
        decoration: InputDecoration(
          isDense: true,
          hintText: hint,
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 4, vertical: 6),
          filled: !plain,
          border: plain ? InputBorder.none : null,
          enabledBorder: plain ? InputBorder.none : null,
          focusedBorder: plain ? InputBorder.none : null,
        ),
      ),
    );
  }

  Widget _doneButton(BuildContext context, bool completed, Color success) {
    return GestureDetector(
      onTap: onToggle,
      child: Container(
        width: 28,
        height: 28,
        decoration: BoxDecoration(
          color: completed ? success : Colors.transparent,
          border: Border.all(
              color: completed ? success : context.appColors.border),
          borderRadius: BorderRadius.circular(8),
        ),
        child: Icon(
          Icons.check,
          size: 16,
          color: completed ? Colors.white : context.appColors.mutedForeground,
        ),
      ),
    );
  }
}
